New Vaadin UI integration

MartinaeyNL opened the pull request:
## Description
This PR replaces almost all of the `<or-mwc-input>` components with the new `<or-vaadin-*>` design system.
It's a very large one, with a lot of changes, that I'll try to summarize in the changelog.
Feel free to test out this branch, and see the changes for yourself.

## Changelog
### Migrating Material UI to Vaadin
- Migrated most of the input fields from Material UI to [Vaadin Web Components](https://vaadin.com/docs/latest/components).
This is done by hand, where all Material components are changed to the new `<or-vaadin-*>` format.
Most of the HTML attributes are similar, or exactly the same, as Vaadin follows the common Web API specification.
It's also common to use slotted elements now, for example displaying labels with `<or-translate>`.
In my personal experience, the syntax is more professional and easier to understand.
Fixes #2593 
- Deprecated the old `<or-mwc-*>` components, as they'll be unmaintained after merging this.
- Bump the Vaadin components dependency to the latest version, `v25.1.3`
### Visual changes
- Replaced the default Manager UI font from system defaults (like Segoe UI) to use [Inter](https://fonts.google.com/specimen/Inter) for every operating system.
The typefont files are now bundled in `@openremote/theme` using RSPack.
For the Manager UI, it acts similar to how we handle and bundle images.
### Bugfixes
- Improved input validity checks across several input fields, to prevent malicious input.
- Improved Rules editor when the user only has READ access.
Fixes #2682
- Bugfix in Chart and Bar Chart widgets, where time presets could be selected while a custom timeframe was set.
- Data export page is now only accessible for users with `READ_ASSETS` role.
### Other
- Minor adjustments of app and component tests, removing hardcoded `<or-mwc-*>`, and simplifying the syntax.
- Added experimental RSBuild tasks in the `package.json`, to allow testing with RSBuild tooling.
We already use it for Storybook, but might for all apps in the foreseeable future. Unstable as of now.
- Cleanup of unused UI code.

## Checklist
- [ ] 1. Acceptance criteria of the linked issue(s) are met
- [ ] 2. Tests are written and all tests pass
- [ ] 3. Changes are manually tested by you and the reviewer
- [ ] 4. Documentation is written or updated

MartinaeyNL committed:
[UI Preview] Add initial set of or-vaadin-components, applied them to Assets page (#2340)

MartinaeyNL committed:
Added default theme colors for dark- and light mode.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix for tsconfig not being correctly formatted after resolving merge conflicts

Ekhorn commented on ui/component/or-vaadin-components/src/or-vaadin-input.ts:
Small detail but a readonly input can still be clicked to put in the caret (it shouldn't be interactive whatsoever in my opinion).
<img width="193" height="74" alt="image" src="https://github.com/user-attachments/assets/06ff5d69-896e-4dba-8f0d-696133e2a760" />

Ekhorn replied:
Only on Firefox lol

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Added Vaadin dialog

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

wborn committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Resolve conflicts

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch master into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix for compile issues

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Add select docs. Attribute history attribute selector now uses Vaadin select menu

MartinaeyNL committed:
Attribute history now uses Vaadin

MartinaeyNL committed:
Change or-asset-viewer header to Vaadin

MartinaeyNL committed:
Migrate several attribute buttons to Vaadin

MartinaeyNL committed:
More attribute field migrations to Vaadin

MartinaeyNL committed:
Replace Material with Vaadin in asset tree

MartinaeyNL committed:
Add combo box, and apply to asset tree.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

Ekhorn committed:
Add missing project reference to tsconfig files

MartinaeyNL committed:
Add list box. Fix for UI test failing

MartinaeyNL committed:
Add @vaadin/item dependency to package.json

MartinaeyNL committed:
Fix yarn lock

MartinaeyNL committed:
Fix for asset viewer test with new invalidation behavior.

MartinaeyNL committed:
Remove accidental ui suffix in test command

MartinaeyNL committed:
Fix select controls for attributes in or-attribute-input. Location field now uses Vaadin as well.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Dependency fix for or-map

MartinaeyNL committed:
Fix or-map dependencies

MartinaeyNL committed:
Fix for test referencing old or-mwc-input, now using or-icon instead

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Add asset dialog now uses vaadin-list instead of mwc-list

MartinaeyNL committed:
WIP

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
WIP

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Add slider component. Upgrade to Vaadin 25.1.0 BETA 1

MartinaeyNL committed:
Fix 'add asset' test with latest changes

MartinaeyNL committed:
Fix gateway asset name in test

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL commented:
I updated this branch, and fixed the merge conflicts within the GitHub UI, but might have made mistakes there.
I will confirm this when I continue on this PR again.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix conflict

MartinaeyNL committed:
Upgrade to Vaadin 25.1.0. Add or-vaadin-menu-bar component. Asset card now uses Vaadin. Fix and migration for sort button on asset tree

MartinaeyNL committed:
Remove Material from the header

MartinaeyNL committed:
Remove all Material controls from or-asset-tree

MartinaeyNL committed:
Revert pagination change

MartinaeyNL committed:
Remove Material from 'add attribute' panel

MartinaeyNL committed:
Asset import and discovery now use Vaadin

MartinaeyNL committed:
Fix for attribute input not applying changes. Replace Group Asset UI with Vaadin

MartinaeyNL committed:
Replace Material UI in or-attribute-barchart

MartinaeyNL committed:
Remove Material UI inputs from or-attribute-card, along with other unused code. (KPI widget)

MartinaeyNL committed:
Removed Material UI inputs from or-attribute-picker

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
or-chart now uses Vaadin UI inputs. Fix for or-attribute-barchart not disabling correctly

MartinaeyNL committed:
Migrated or-dashboard-builder to new Vaadin components. Lots of other minor changes as well.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge

MartinaeyNL committed:
Attempt to remove Material UI from or-json-forms.

MartinaeyNL committed:
Migrate or-log-viewer to Vaadin components.

MartinaeyNL committed:
Update copy button on logs page as well.

MartinaeyNL committed:
A big one: migrated most of the Rules UI to Vaadin

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix for conflict in getValueHolderInputTemplateProvider. Moved the updated logic from or-mwc-input to or-vaadin-input

MartinaeyNL committed:
Add vaadin components dependency to packages

MartinaeyNL committed:
Fix yarn.lock

MartinaeyNL committed:
Fix tests. Improve test app config file. Improvements to or-vaadin-combo-box items change detection. Realm selector in header now disables correctly.

MartinaeyNL committed:
Fix or-json-forms tests

MartinaeyNL committed:
Header fix for Firefox: replaced alarm icon link with Vaadin button

MartinaeyNL committed:
Migrate account and gateway pages to Vaadin. Add new (custom) or-vaadin-email-field component

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Add Vaadin dependency to Manager app

MartinaeyNL committed:
Yarn lock fix

MartinaeyNL committed:
Migrate alarms to Vaadin

MartinaeyNL committed:
More minor Vaadin UI migrations on page level (assets and appearance page)

MartinaeyNL committed:
Migrate export page to Vaadin

MartinaeyNL committed:
Migrate gateway page to Vaadin

MartinaeyNL committed:
Migrate appearance page to Vaadin. Add 'type' attribute to or-vaadin-text-field

MartinaeyNL committed:
Attempt to fix test that was specifically written for or-mwc-input.

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Temporarily disabled UI app tests

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge with master

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-map/or-conf-map-card.ts:
```diff
@@ -250,43 +250,42 @@ export class OrConfMapCard extends LitElement {
+                                                    }}>
+                                <or-translate slot="label" value="configuration.minZoom"></or-translate>
+                            </or-vaadin-number-field>
+                            <or-vaadin-button theme="icon" @click=${() => this.setZoom(map.minZoom)}>
+                                <or-icon icon="eye" style="margin: 0;"></or-icon>
+                            </or-vaadin-button>
```
Might need to discuss how we deal with icons on inputs.

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-map/or-conf-map-geojson.ts:
```diff
@@ -22,7 +22,12 @@ export class OrConfMapGeoJson extends LitElement {
     protected render() {
-        return html`<or-mwc-input type="${InputType.BUTTON}" label="geoJson" outlined icon="pencil" @or-mwc-input-changed="${this.showDialog}"></or-mwc-input>`
+        return html`
+            <or-vaadin-button @click=${() => this.showDialog()}>
+                <or-icon slot="prefix" icon="pencil"></or-icon>
+                <or-translate value="configuration.geoJson"></or-translate>
+            </or-vaadin-button>
```
Also good to discuss if we want to integrate or-translate by default.

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-map/or-conf-map-global.ts:
```diff
@@ -108,17 +108,16 @@ export class OrConfMapGlobal extends LitElement {
-                            this.notifyConfigChange(this.config);
-                        }}"
-                    ></or-mwc-input>
+                    <or-vaadin-text-field class="input" type="url" pattern="^[a-zA-Z][a-zA-Z0-9+.-]*://.*$"
```
I find it weird we would need to define the pattern for a URL when you specify the type. If possible use the Vaadin default or move the pattern to the component itself.

MartinaeyNL replied:
`type="url"` is mostly meant for accessibility reasons.
The type specification doesn't include any pattern checking, therefore I've added it myself;
https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/url

Vaadin doesn't have any documentation on using URLs in their inputs.
Which basically tells us they rely on the native `<input>` implementation haha

Btw, originally, the `type` attribute is not supported on `<vaadin-text-field>`, so I've added it myself;
```
    /**
     * Custom `type` attribute that echoes to the input element.
     * This is useful for types like 'url' and 'email'.
     * Use with caution: using unknown types, or mismatched types like 'number' could potentially break the component.
     */
    @property({type: String})
    public type?: string;
```

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-map/or-conf-map-global.ts:
```diff
@@ -152,17 +151,15 @@ export class OrConfMapGlobal extends LitElement {
-                                this.requestUpdate();
-                            }}"
-                        ></or-mwc-input>
+                        <or-vaadin-text-field class="input" type="url" pattern="^[a-zA-Z][a-zA-Z0-9+.-]*://.*$"
```
https://github.com/openremote/openremote/pull/2377/changes#r3316750561

MartinaeyNL committed:
Fix for asset public read toggle not working

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
-                                      onchange?.(user, true, this._isInvalid());
-                                  }}"
-                    ></or-mwc-input>
+                    <or-vaadin-text-field id="new-username" class="validate" disabled
```
Why is the h5 removed here?


```suggestion
                    <h5>${i18next.t("details")}</h5>
                    <or-vaadin-text-field id="new-username" class="validate" disabled
```

MartinaeyNL replied:
Good spot! I did remove it because I thought it was ugly 😂 
But I kept it on the other pages, so I'll put it back 👍

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
+                                           ?required=${registrationEmailAsUsername}
+                                           ?readonly=${(!!user?.id && registrationEmailAsUsername) || readonly}
+                                           ?disabled="${!user || (!!user?.id && registrationEmailAsUsername)}"
+                                           pattern="^[\\w\\.\\-]+@([\\w\\-]+\\.)+[\\w]{2,4}$"
```
Again unless we have a reason to set our own pattern the field should have 1 by default.

MartinaeyNL replied:
I've added a default pattern to `<or-vaadin-email-field>`.
But to prevent any issues during migrating to Vaadin, I've simply copy pasted the pattern here.
For example, maybe Keycloak uses the pattern specified here.
```
constructor() {
        super();

        // Instead of using the `@vaadin/email-field` package, we use this _setType() method.
        // This is because Vaadin their package, directly inherits TextField, causing the component to break.
        // Therefore, we manually set the input type and pattern, which is not any different from the Web API specification.
        (this as any)._setType('email');
        this.pattern = '^[a-zA-Z0-9_\\-+]+(?:\\.[a-zA-Z0-9_\\-+]+)*@[a-zA-Z0-9\\-.]+\\.[a-zA-Z0-9\\-]{2,}$';
    }
```

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
+                    </or-vaadin-email-field>
+                    <or-vaadin-text-field id="new-firstName" class="validate"
+                                          ?readonly=${readonly} ?disabled=${!user}
+                                          minLength="5" autocomplete="false"
```
Why was it 5 before tf?

```suggestion
                                          minLength="1" autocomplete="false"
```

MartinaeyNL replied:
Good spot! I'll put it back.

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
+                    </or-vaadin-text-field>
                 </div>
                 <div class="column">
-                    <h5>${i18next.t("password")}</h5>
```
Why is the h5 removed here?

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
-                                      .disabled="${!user.email}"
-                                      @or-mwc-input-changed="${(e: OrInputChangedEvent) => this._onResetPasswordBtnClick(e)}"
-                        ></or-mwc-input>
+                        <or-vaadin-button id="reset-password" theme="primary" ?disabled=${!user.email} style="width: fit-content;"
```
A possible reason to adjust the default button styling?

```suggestion
                        <or-vaadin-button id="reset-password" theme="primary" ?disabled=${!user.email}
```

MartinaeyNL replied:
It was one of the few breaking changes in migrating from Material to Vaadin.
In Material, every button was using the minimal width it needed, unless specified differently.
In Vaadin, every button fills the available width, unless specified differently.
So I needed to use `width: fit-content;` here to make it smaller 😉

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -248,95 +247,79 @@ export class PageAccount extends Page<AppStateKeyed> {
-                                          onchange?.(user, changed, this._isInvalid());
-                                      }}"
-                        ></or-mwc-input>
+                        <or-vaadin-password-field id="new-password" class="validate"
```
The default vaadin password field is not very strict. We might want to consider adding a strict option to the field with similar conditions as shown on https://vaadin.com/docs/latest/components/password-field#validation

MartinaeyNL replied:
I agree with you here.
But unfortunately, as the password policy is specified on Keycloak level, we shouldn't "restrict" users here.
Instead, the `_passwordPolicy` (aka the requirements) is shown in text below the password fields.
If the user fills in an invalid password, the HTTP API will throw an error.

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -369,22 +350,30 @@ export class PageAccount extends Page<AppStateKeyed> {
-                this._repeatPasswordElem?.setCustomValidity(error);
+                console.warn("Could not update password: the passwords do not match.");
                 user.password = "";
+                setTimeout(() => {
```
reminder @Ekhorn  test these changes

Ekhorn commented on ui/app/manager/src/pages/page-account.ts:
```diff
@@ -423,7 +412,9 @@ export class PageAccount extends Page<AppStateKeyed> {
-        return Array.from(validateArray).filter(e => e instanceof OrMwcInput).some(input => !(input as OrMwcInput).valid);
+        return Array.from(validateArray)
+            .filter(e => e instanceof HTMLInputElement)
+            .some(input => !(input as HTMLInputElement).checkValidity());
```
Something like this might work?

```suggestion
            .some(input: input is HTMLInputElement => !(input as HTMLInputElement).checkValidity());
```

wborn commented:
@codex review

Ekhorn commented on ui/app/manager/src/pages/page-configuration.ts:
```diff
@@ -285,11 +285,11 @@ export class PageConfiguration extends Page<AppStateKeyed> {
-                                            this.managerConfiguration = val;
-                                        }).catch(e => console.error(e))}"
-                                    ></or-mwc-input>
+                                    <or-vaadin-button style="width: fit-content;" @click=${() => this.getManagerConfig()
```
Again might want to consider changing the default or-vaadin-button styling.

Ekhorn commented on ui/app/manager/src/pages/page-export.ts:
check @Ekhorn

Ekhorn commented on ui/app/manager/src/pages/page-realms.ts:
Come back to this

Ekhorn commented on ui/app/manager/test/fixtures/pages/realms.ts:
```diff
@@ -13,15 +13,18 @@ export class RealmsPage implements BasePage {
+      await Promise.all([
+        this.page.waitForLoadState('networkidle'), // Wait for reload to take place after adding realm
+        this.page.getByRole("button", { name: "create" }).click()
+      ]);
```
It is bad practice to wait for networkidle, it is better to wait for a certain request path using `waitForURL`.

```suggestion
      await page.waitForURL("**/realm/**");
      await this.page.getByRole("button", { name: "create" }).click()
```

Ekhorn commented on ui/app/manager/test/fixtures/pages/realms.ts:
```diff
@@ -34,5 +37,6 @@ export class RealmsPage implements BasePage {
     await this.page.getByRole("button", { name: "Delete" }).click();
     await this.page.getByRole("alertdialog").getByRole("textbox", { name: "Realm" }).fill(realm);
     await this.page.getByRole("button", { name: "OK" }).click();
+    await this.page.waitForLoadState('networkidle') // Wait for reload to take place after deleting realm
```
```suggestion
```

Ekhorn commented on ui/app/manager/test/insights.test.ts:
```diff
@@ -105,16 +105,16 @@ test("Create a Map widget with text thresholds", async ({ manager, shared, page,
 
     // Check default text configuration is correct
-    const textThresholds = mwcInput.getInputByType("text", thresholdPanel);
+    const textThresholds = thresholdPanel.locator("input[type='text']"); // mwcInput.getInputByType("text", thresholdPanel);
```
What happened here lol

Ekhorn commented on ui/app/manager/test/rules.test.ts:
```diff
@@ -109,25 +109,29 @@ test("Create a When-Then rule for an asset with a trigger and action", async ({p
+    // When clause
+    const when = page.locator("or-rule-when");
+    await when.getByRole("button", { name: "Add condition", exact: true }).click();
+    await when.locator("li[role='menuitem']").filter({hasText: energyRule.asset_type}).click();
```
Guessing this can't be role based?

MartinaeyNL replied:
Couldn't find a quick fix for it. So I left it as is.

Ekhorn commented on ui/app/manager/rsbuild.config.ts:
Eh, what was the idea of this again? Is rsbuild with this config now required in development? I really want these alias issues to be fixed before we adopt his.

MartinaeyNL replied:
No, it was mainly for testing / experimental purposes.
We added an `experimental` option in the `package.json` file, that uses RSBuild instead of RSPack.
So fully optional in case you want to test out RSBuild compability down the road once we make improvements.
```json
"serve": "npx tsc -b --clean && npx shx rm -rf dist lib && npx orutil build && npx cross-env NODE_OPTIONS=--max_old_space_size=4096 rspack serve --mode development",
"serve:experimental": "npx tsc -b --clean && npx shx rm -rf dist lib && npx orutil build && npx cross-env NODE_OPTIONS=--max_old_space_size=4096 rsbuild dev",
```

Ekhorn commented on ui/component/core/src/defaults.ts:
Have we started using these yet?

MartinaeyNL replied:
Good spot! Not the ones I have added no.
As we introduced the `@openremote/theme` package, I can remove my changes here.
Because every styling change in this PR has been using CSS only, without any TypeScript references 👍
I'll remove these.

Ekhorn commented on ui/component/or-app/src/or-header.ts:
Maybe need a bit of an explanation.

MartinaeyNL committed:
Migrate users page. Update vaadin dependency. Styling fix for or-vaadin-checkbox.

Ekhorn commented on ui/component/or-asset-tree/src/or-add-asset-dialog.ts:
```diff
@@ -301,17 +289,30 @@ export class OrAddAssetDialog extends LitElement {
             attributes: AttributeDescriptor[] | undefined = assetTypeInfo?.attributeDescriptors?.filter(e => !e.optional),
             optionalAttributes: AttributeDescriptor[] | undefined = assetTypeInfo?.attributeDescriptors?.filter(e => !!e.optional);
 
+        console.debug("Rendering", this.selectedAttributes);
```
```suggestion
```

Ekhorn commented on ui/component/or-asset-tree/src/or-add-asset-dialog.ts:
```diff
@@ -355,13 +361,14 @@ export class OrAddAssetDialog extends LitElement {
     protected async onTypeChanged(isAgent: boolean, listItem: ListItem) {
         await this.updateComplete;
 
+        console.debug(this.selectedAttributes);
```
```suggestion
```

Ekhorn commented on ui/component/or-asset-viewer/src/index.ts:
```diff
@@ -733,11 +753,16 @@ function getPanelContent(id: string, assetInfo: AssetInfo, hostElement: LitEleme
+                                                    if((ev.currentTarget as OrVaadinCheckbox).checked) {
+                                                        newlySelectedAttributes.push(attribute);
+                                                    } else {
+                                                        newlySelectedAttributes.splice(newlySelectedAttributes.findIndex((s) => s === attribute), 1);
```
```suggestion
                                                        newlySelectedAttributes.splice(newlySelectedAttributes.indexOf(attribute), 1);
```

MartinaeyNL committed:
Migrate appearance page realm config to Vaadin. Bump Vaadin lumo styles dependency. Minor cleanup

MartinaeyNL committed:
Fix header menu on mobile

Ekhorn reviewed (commented):
I really like the improved locators in the tests. This is about 1/4 will continue. I'll see if I can get to the JSON Forms migration this sprint or not.

Ekhorn commented on ui/component/or-asset-viewer/src/index.ts:
```diff
@@ -1446,6 +1494,7 @@ export class OrAssetViewer extends subscribe(manager)(translate(i18next)(LitElem
         }
 
         const asset = this._assetInfo.asset;
+        console.debug("Asset to be saved:", asset);
```
Still needed?

MartinaeyNL committed:
Migrate realms page to Vaadin

MartinaeyNL committed:
Migrate Roles page to Vaadin

MartinaeyNL committed:
Migrate provisioning page to Vaadin

Ekhorn commented on ui/component/or-asset-viewer/src/or-edit-asset-panel.ts:
```diff
@@ -409,6 +415,7 @@ export class OrEditAssetPanel extends LitElement {
     }
 
     protected _onAttributeModified(attribute: Attribute<any>, newValue: any) {
+        console.debug("_onAttributeModified", attribute, newValue);
```
Still needed?

```suggestion
```

Ekhorn commented on ui/component/or-asset-viewer/src/or-edit-asset-panel.ts:
```diff
@@ -441,14 +448,14 @@ export class OrEditAssetPanel extends LitElement {
         if (!valueDescriptor) {
             console.log("Couldn't find value descriptor for meta item so falling back to simple JSON input: " + metaItem.name);
-            content = html`<or-mwc-input @or-mwc-input-changed="${(ev: OrInputChangedEvent) => this._onMetaItemModified(attribute, metaItem, ev.detail)}" .type="${InputType.JSON}" .value="${metaItem.value}"></or-mwc-input>`;
+            content = html`<or-mwc-input @or-mwc-input-changed="${(ev: OrInputChangedEvent) => this._onMetaItemModified(attribute, metaItem, ev.detail.value)}" .type="${InputType.JSON}" .value="${metaItem.value}"></or-mwc-input>`;
```
Not migrated yet.

MartinaeyNL replied:
Yeah I don't have a JSON editor for Vaadin yet. I guess it's part of the JSON forms issue on Vaadin in #2722

MartinaeyNL committed:
Fix expand all toggle in asset viewer. Added logging for asset modification to debug

Ekhorn commented on ui/component/or-attribute-input/src/index.ts:
```diff
@@ -581,6 +574,7 @@ export class OrAttributeInput extends subscribe(manager)(translate(i18next)(LitE
         };
 
         if (this.customProvider) {
+            console.debug("Rendering a custom or-attribute-input provider...");
```
Still needed?

```suggestion
```

MartinaeyNL committed:
Address review comments page-account.ts

Ekhorn commented on ui/component/or-attribute-history/src/index.ts:
```diff
@@ -891,5 +903,11 @@ export class OrAttributeHistory extends translate(i18next)(LitElement) {
         });
     }
+
+    protected _getDefaultDate() {
+        const date = new Date();
+        date.setHours(date.getHours() + 1, 0, 0, 0); // Round to the end of the hour
+        return date;
+    }
```
```suggestion
    protected _getDefaultDate() {
        const date = new Date();
        date.setHours(date.getHours() + 1, 0, 0, 0); // Round to the beginning of the next hour
        return date;
    }
```

MartinaeyNL committed:
Remove console statements or-add-asset-dialog

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

Ekhorn commented on ui/component/or-attribute-picker/src/asset-attribute-picker.ts:
```diff
@@ -186,6 +192,7 @@ export class OrAssetAttributePicker extends AttributePicker {
      * So, selected attributes of other assets, will merge together with the new {@link attrNames}.
      */
     protected _onAttributesSelect(attrNames: string[]) {
+        console.debug(attrNames);
```
```suggestion
```

MartinaeyNL committed:
Fix translation

Ekhorn commented on ui/component/or-attribute-picker/src/assettype-attribute-picker.ts:
```diff
@@ -196,6 +212,7 @@ export class OrAssetTypeAttributePicker extends AttributePicker {
      * HTML Callback function when the selected asset type has been updated.
      */
     protected _onAssetTypeItemClick(listItem: ListItem) {
+        console.debug(listItem);
```
```suggestion
```

Ekhorn commented on ui/component/or-dashboard-builder/src/settings/chart-settings.ts:
```diff
@@ -536,16 +562,20 @@ export class ChartSettings extends WidgetSettings {
 
-    protected onSamplingQueryChange(ev: OrInputChangedEvent) {
-        this.widgetConfig.datapointQuery.type = this.samplingOptions.get(ev.detail.value)! as any;
+    protected onSamplingQueryChange(ev: Event) {
```
Why is not using `samplingOptions` anymore?

MartinaeyNL replied:
Since the new `<or-vaadin-select>` allows to inject full JSON objects (like `this.samplingOptions`).
Like so:
```html
<or-vaadin-select .items=${samplingOptions} value=${samplingValue} @change=${(ev: Event) => this.onSamplingQueryChange(ev)}>
```
I simply have to call `ev.currentTarget.value`, and it will return the object with type safety. 👍

Ekhorn commented on ui/component/or-dashboard-builder/src/settings/table-settings.ts:
```diff
@@ -87,9 +106,11 @@ export class TableSettings extends AssetWidgetSettings {
+        const elem = ev.currentTarget as OrVaadinSelect;
+        const value = Number(elem.value || 10);
         this.widgetConfig.tableSize = value;
+        console.debug(this.widgetConfig.tableSize);
```
```suggestion
```

Ekhorn commented on ui/component/or-dashboard-builder/src/settings/table-settings.ts:
```diff
@@ -87,9 +106,11 @@ export class TableSettings extends AssetWidgetSettings {
-    protected onTableSizeSelect(ev: OrInputChangedEvent) {
-        const value = ev.detail.value || 10;
+    protected onTableSizeSelect(ev: Event) {
+        const elem = ev.currentTarget as OrVaadinSelect;
+        const value = Number(elem.value || 10);
         this.widgetConfig.tableSize = value;
+        console.debug(this.widgetConfig.tableSize);
         if(value !== 10) {
```
See https://github.com/openremote/openremote/pull/2377/changes#r3324140581

Ekhorn commented on ui/component/or-dashboard-builder/src/index.ts:
```diff
@@ -326,6 +343,7 @@ export class OrDashboardBuilder extends translate(i18next)(LitElement) {
     }
 
     willUpdate(changedProps: PropertyValues) {
+        console.debug(changedProps);
```
```suggestion
```

Ekhorn commented on ui/component/or-dashboard-builder/src/index.ts:
```diff
@@ -586,6 +607,7 @@ export class OrDashboardBuilder extends translate(i18next)(LitElement) {
             display: (this.editMode && (this._isReadonly() || !this._hasEditAccess())) ? 'none' : undefined,
             maxHeight: this.editMode ? "calc(100vh - 77px - 50px)" : "inherit"
         };
+        console.debug(`Rendering dashboard UI with loading = ${this.isLoading} and hasChanged = ${this.hasChanged}`);
```
Still needed?
```suggestion
```

Ekhorn commented on ui/component/or-dashboard-builder/src/or-dashboard-boardsettings.ts:
```diff
@@ -72,13 +89,9 @@ export class OrDashboardBoardsettings extends translate(i18next)(LitElement) {
                             <div class="label">
                                 ${html`<span>${unsafeHTML(i18next.t('dashboard.whoCanView').toString())}</span>`}
```
```suggestion
<or-translate value="dashboard.whoCanView"></or-translate>
```

MartinaeyNL replied:
Before, `unsafeHTML()` is used for allowing **bold text** inside the translation files.
But I removed it now, as I think it is overcomplicating things. 👍

Ekhorn commented on ui/component/or-dashboard-builder/src/or-dashboard-boardsettings.ts:
```diff
@@ -177,23 +188,21 @@ export class OrDashboardBoardsettings extends translate(i18next)(LitElement) {
                 return html`
                     <div style="margin-bottom: ${screenPresets.length > 1 ? '24px' : '16px'}">
                         <div class="label">
                             ${html`<span>${unsafeHTML(i18next.t("dashboard.onScreenMyBoardShould").replace("{{size}}", ("<b>" + i18next.t(preset.displayName!) + "</b>")))}</span>`}
```
This can be refactored won't attempt a suggestion lol

Ekhorn commented on ui/component/or-dashboard-builder/src/or-dashboard-tree.ts:
```diff
@@ -224,34 +243,30 @@ export class OrDashboardTree extends translate(i18next)(LitElement) {
-                                ></or-mwc-input>
-                                <or-mwc-input type="${InputType.BUTTON}" icon="delete" title="${i18next.t("delete")}" @or-mwc-input-changed="${() => {
+                        ${when(this.selected != null, () => html`
+                            <or-vaadin-button theme="icon" title=${i18next.t("deselect")} @click=${() => this.selectDashboard(undefined)}>
```
Can't this be a `<or-translate value="deselect">`?

MartinaeyNL replied:
No because it is using the HTML native `title="xxx"` to generate a tooltip.
For simplicity I kept it there as before.

Ekhorn replied:
I think we could integrate localization for displayable text attributes into the vaadin components, but I guess that may be a separate issue.

Ekhorn commented on ui/component/or-dashboard-builder/src/or-dashboard-tree.ts:
```diff
@@ -224,34 +243,30 @@ export class OrDashboardTree extends translate(i18next)(LitElement) {
+                            `)}
+                        `)}
+                        ${when(!this.readonly, () => html`
+                            <or-vaadin-button theme="icon" class="hideMobile" title=${i18next.t("addInsights")} @click=${() => this.createDashboard(DashboardSizeOption.DESKTOP)}>
```
Same thing here with or-translate

MartinaeyNL replied:
Same reply, as it was using `title="xxx"` for generating a tooltip, not a label inside the button.

Ekhorn commented on ui/component/or-json-forms/src/layouts/layout-vertical-element.ts:
```diff
@@ -253,7 +265,12 @@ export class LayoutVerticalElement extends LayoutBaseElement<VerticalLayout | Gr
+                                                          const elem = ev.currentTarget as OrVaadinTextField;
+                                                          keyChangeHandler(elem, key, elem.value)
+                                                      }}>
+                                </or-vaadin-text-field>>
```
```suggestion
                                </or-vaadin-text-field>
```

Ekhorn commented on ui/component/or-json-forms/src/layouts/layout-vertical-element.ts:
```diff
@@ -253,7 +265,12 @@ export class LayoutVerticalElement extends LayoutBaseElement<VerticalLayout | Gr
                         <div class="row">
                             <div class="key-container">
-                                <or-mwc-input .type="${InputType.TEXT}" @or-mwc-input-changed="${(ev:OrInputChangedEvent) => keyChangeHandler(ev.currentTarget as OrMwcInput, key, ev.detail.value)}" required .pattern="${dynamicPropertyRegex}" .value="${key}"></or-mwc-input>
+                                <or-vaadin-text-field value=${key} required pattern="${dynamicPropertyRegex}" 
+                                                      @change=${(ev: Event) => {
+                                                          const elem = ev.currentTarget as OrVaadinTextField;
+                                                          keyChangeHandler(elem, key, elem.value)
+                                                      }}>
+                                </or-vaadin-text-field>>
                             </div>
```
https://github.com/openremote/openremote/pull/2377/changes#r3324287401

Ekhorn commented on ui/component/or-mwc-components/src/or-mwc-list.ts:
```diff
@@ -61,6 +61,14 @@ export type ListGroupItem = {
+     *                 ${list.list}
+     *             `)}
+     *         </or-vaadin-list-box>
+     */
```
```suggestion
```

Ekhorn commented on ui/component/or-rules/src/json-viewer/forms/or-rule-form-alarm.ts:
translation key path changed?

MartinaeyNL replied:
Yes that is correct. I saw the label was simply not localized yet, while there was a translation key in the files.
So I've updated those, and all work as expected now. _The bug has been fixed_ 😂

Ekhorn commented on ui/component/or-rules/src/json-viewer/or-rule-asset-query.ts:
```diff
@@ -341,23 +352,27 @@ export class OrRuleAssetQuery extends translate(i18next)(LitElement) {
-                <or-mwc-input type="${InputType.BUTTON}" .readonly="${this.readonly || false}" icon="clock-plus-outline"
-                              title="${i18next.t("rulesEditorAddDuration")}" @or-mwc-input-changed="${() => onAdd(index)}"
-                ></or-mwc-input>
+                <or-vaadin-button theme="icon" ?disabled=${this.readonly} title=${i18next.t("rulesEditorAddDuration")}
```
or-translate?

MartinaeyNL replied:
`title="xxx"` is meant for the native HTML tooltips you get on elements.
There are better solution to this, but for now I kept the behavior the same as before.

Ekhorn commented on ui/component/or-rules/src/json-viewer/or-rule-trigger-query.ts:
```diff
@@ -194,6 +210,7 @@ export class OrRuleTriggerQuery extends LitElement {
             const date = new Date();
             date.setHours(Number(splittedTime[0]));
             date.setMinutes(Number(splittedTime[1]));
+            console.debug(date.toString());
```
```suggestion
```

Ekhorn commented on ui/component/or-rules/src/json-viewer/or-rule-trigger-query.ts:
```diff
@@ -194,6 +210,7 @@ export class OrRuleTriggerQuery extends LitElement {
             const date = new Date();
             date.setHours(Number(splittedTime[0]));
             date.setMinutes(Number(splittedTime[1]));
+            console.debug(date.toString());
             this.condition.cron = Util.formatCronString(undefined, undefined, undefined, date.getUTCHours().toString(), date.getUTCMinutes().toString());
             this.dispatchEvent(new OrRulesJsonRuleChangedEvent());
```
https://github.com/openremote/openremote/pull/2377/changes#r3324953156

MartinaeyNL committed:
Enable tests again. Fix for several tests. Other cleanup

Ekhorn commented on ui/component/or-vaadin-components/src/or-vaadin-input.ts:
```diff
@@ -32,6 +35,7 @@ import "./or-vaadin-text-area";
  * @param constructors - List of CustomElements to scan for available properties
  */
 function registerProperties(constructors: (CustomElementConstructor | undefined)[]) {
+    console.debug("[OrVaadinInput] Registering properties for", constructors);
```
```suggestion
```

Ekhorn commented on ui/component/or-vaadin-components/src/or-vaadin-item.ts:
```diff
@@ -0,0 +1,27 @@
+import {OrVaadinComponent} from "./util";
+import {type LitElement} from "lit";
+
+@customElement("or-vaadin-item")
+export class OrVaadinListBox extends (Item as new () => Item & LitElement) implements OrVaadinComponent {
+}
```
Agree with this.

Ekhorn replied:
https://github.com/openremote/openremote/pull/2377/changes#r3325004496

Ekhorn commented on ui/component/or-vaadin-components/src/or-vaadin-item.ts:
```diff
@@ -0,0 +1,27 @@
+import {type LitElement} from "lit";
+
+@customElement("or-vaadin-item")
+export class OrVaadinListBox extends (Item as new () => Item & LitElement) implements OrVaadinComponent {
```
```suggestion
export class OrVaadinItem extends (Item as new () => Item & LitElement) implements OrVaadinComponent {
```

MartinaeyNL committed:
Few more test fixes

Ekhorn reviewed (commented):
2/3 ;)

MartinaeyNL committed:
Revert accidental npm test modification

Ekhorn reviewed (commented):
Ok, that's all! Will need to do manual testing, but happy with the reduction of the styles and better locators in test files. I will see what we can do on the JSON Forms.

Nice work!

MartinaeyNL committed:
More UI test fixes

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Cleanup

MartinaeyNL committed:
Cleanup

MartinaeyNL committed:
Update ui/component/or-asset-viewer/src/index.ts

Co-authored-by: Koen <44117179+Ekhorn@users.noreply.github.com>

MartinaeyNL committed:
Fix tests

MartinaeyNL committed:
Update ui/component/or-attribute-history/src/index.ts

Co-authored-by: Koen <44117179+Ekhorn@users.noreply.github.com>

MartinaeyNL committed:
Added vaadin select in 'add realm config' dialog. Minor cleanup.

MartinaeyNL committed:
Migrate logs category picker to Vaadin. Merge with master.

MartinaeyNL committed:
Added Inter font, and applied to the default theme. Manager app doesn't have their own font anymore.

MartinaeyNL committed:
Fix for checkbox text not being clickable due to use of or-translate.

MartinaeyNL committed:
Fix hr styling for Firefox

MartinaeyNL committed:
Fix menu bar button. Made hr element less prominent.

MartinaeyNL committed:
Unused dependencies fix

MartinaeyNL committed:
Temporary log scan report

MartinaeyNL committed:
Fix for switch test after font change

MartinaeyNL committed:
Bump RSBuild and RSPack to the latest version of this major. Now using @rspack/core v1.7.10

MartinaeyNL committed:
Bump rollup version for addressing vulnerability

MartinaeyNL committed:
Bump vite version

MartinaeyNL committed:
Bump shell-quote version

MartinaeyNL committed:
Merge with master, and fix yarn.lock conflict

MartinaeyNL committed:
Revert yarn.lock version downgrade

MartinaeyNL commented:
I'm updating the changelog to mirror the changes I've made here, as there are a lot of them.
Hopefully that will properly explain what this PR includes 😉 

Besides that, this PR is ready for a final review.
I have deployed the latest changes to a test server of ours, so please take a look.
Let me know if I could help with anything.

MartinaeyNL committed:
Fix for ENTER not working in attribute inputs

MartinaeyNL committed:
Update styling of panel title

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix for HTML tag for filtering alarm status not being correct.

MartinaeyNL committed:
Merge with master. Resolved some review comments.

wborn commented:
@codex review

Ekhorn commented on ui/app/manager/test/fixtures/pages/realms.ts:
```diff
@@ -39,5 +39,6 @@ export class RealmsPage implements BasePage {
     await this.page.getByRole("button", { name: "Delete" }).click();
     await this.page.getByRole("alertdialog").getByRole("textbox", { name: "Realm" }).fill(realm);
     await this.page.getByRole("button", { name: "OK" }).click();
+    // await this.page.waitForLoadState('networkidle') // Wait for reload to take place after deleting realm
```
`networkidle` is bad practice remove it completely.

```suggestion
```

MartinaeyNL replied:
Ah yes. I kept it "commented" to see if the tests succeed. And they do 😉 
I'll remove this.

Ekhorn commented on ui/component/or-log-viewer/src/index.ts:
Unused `or-mwc-input` imports?

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-json.ts:
Unused `or-mwc-input` import

Ekhorn commented on ui/app/manager/src/components/configuration/or-conf-map/or-conf-map-geojson.ts:
Unused `or-mwc-input` import

Ekhorn commented on ui/component/or-rules/src/or-rule-viewer.ts:
Unused `or-mwc-input` import

Ekhorn commented on ui/component/or-rules/src/json-viewer/or-rule-asset-query.ts:
Unused `or-mwc-input` import

Ekhorn commented on ui/component/or-rules/src/or-rule-group-viewer.ts:
Unused `or-mwc-input` import

Ekhorn reviewed (commented):
Some unused imports left.

Ekhorn commented on ui/app/manager/src/pages/page-realms.ts:
```diff
@@ -324,27 +335,49 @@ export class PageRealms extends Page<AppStateKeyed> {
+                                          </or-vaadin-button>
                                       ` : ``}
-                                      <or-mwc-input style="margin-left: auto;" label="save" .type="${InputType.BUTTON}" @or-mwc-input-changed="${() => this._updateRealm(realm)}"></or-mwc-input>   
+                                      <or-vaadin-button theme="primary" @click=${() => this._updateRealm(realm)}>
```
Reinstate this because it is weird for the user to have the save button suddenly on the other side.

```suggestion
                                      <or-vaadin-button style="margin-left: auto" theme="primary" @click=${() => this._updateRealm(realm)}>
```

Ekhorn commented:
The available pages have increased for users with no permissions (that shouldn't be the case IMO). I can now see all pages in the drop down menu, while before I could only see language, data export and logout.

When I click those pages obviously things don't load. I see completely empty pages, pages with errors and e.g. the appearance page where it looks like I can edit everything while I can't.

MartinaeyNL committed:
Address several review comments with bugs or cleanup steps.

MartinaeyNL committed:
Fix for header menu items not taking roles into account.

MartinaeyNL committed:
Yarn lock fix

MartinaeyNL committed:
Cleanup

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Restrict data export page to users with READ_ASSETS only.

MartinaeyNL committed:
Merge with master

MartinaeyNL committed:
Fix for menu not working on mobile.

MartinaeyNL committed:
Styling fix mobile header

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Fix for #2682 on readonly rules viewer

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

MartinaeyNL committed:
Realm picker and page name is now smaller on mobile

MartinaeyNL committed:
Merge with master

MartinaeyNL committed:
Merge branch 'master' into feature/ui-preview

wborn committed:
Merge branch 'master' into feature/ui-preview

Ekhorn reviewed (approved):
The UI looks much better 😄

@MartinaeyNL amazing work!
